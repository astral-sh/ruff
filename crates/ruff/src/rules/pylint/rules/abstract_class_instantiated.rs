use std::collections::{HashMap, HashSet};

use itertools::Itertools;
use ruff_python_ast::cast;
use ruff_python_semantic::{
    SemanticModel, {Definition, Member, MemberKind},
};
use rustpython_parser::ast::{Expr, ExprName, Ranged, Stmt, StmtClassDef};

use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, violation};

use crate::checkers::ast::Checker;

/// ## What it does
/// Checks that abstract classes with `abc.ABCMeta` as metaclass and abstract methods
/// are not instantiated.
///
/// ## Why is this bad?
/// Abstract classes are not meant to be instantiated, they are meant to be subclassed.
/// Using an instance of an abstract class can lead to unexpected behavior.
///
/// ## Example
/// ```import abc
///
///
/// class Animal(abc.ABC):
///     @abc.abstractmethod
///     def make_sound(self):
///         pass
///
///
/// sheep = Animal()```
///
///
/// ## References
/// - [Python documentation](https://docs.python.org/3/library/abc.html)
/// - [PEP 3119](https://peps.python.org/pep-3119/#the-abc-module-an-abc-support-framework)
#[violation]
pub struct AbstractClassInstantiated {
    class_name: String,
}

impl Violation for AbstractClassInstantiated {
    #[derive_message_formats]
    fn message(&self) -> String {
        format!(
            "Abstract class '{}' with abstract methods instantiated",
            self.class_name
        )
    }
}

const ABC_METHODS: [&str; 4] = [
    "abc.abstractproperty",
    "abc.abstractmethod",
    "abc.abstractclassmethod",
    "abc.abstractstaticmethod",
];

const ABC_METACLASSES: [&str; 2] = ["_py_abc.ABCMeta", "abc.ABCMeta"];

/// Remove typing.Generic from the MRO
fn clean_typing_generic_mro(bases: &mut Vec<String>, bases_mro: &mut Vec<Vec<String>>) {
    if let Some((position_in_bases, _)) = bases
        .iter()
        .find_position(|base| base.as_str() == "typing.Generic")
    {
        if bases_mro
            .iter()
            .enumerate()
            .filter(|&(i, _)| i != position_in_bases)
            .any(|(_, mro)| mro.iter().any(|n| n.as_str() == "typing.Generic"))
        {
            bases.remove(position_in_bases);
            bases_mro.remove(position_in_bases);
        }
    }
}

/// Checks if a MRO has duplicates
fn has_duplicates(mro: &[String]) -> bool {
    let mut seen = HashSet::new();
    for class_def in mro.iter() {
        if seen.contains(class_def) {
            return true;
        }
        seen.insert(class_def);
    }
    false
}

/// Recursively calculates the MRO of a class
fn method_resolution_order(model: &SemanticModel, class_name: String) -> Option<Vec<String>> {
    let mut bases_mro = Vec::new();
    let mut bases = Vec::new();
    let class_definitions = class_definitions(model).collect_vec();
    let Some(class_def) =  class_definitions.iter().find(|class_def| class_def.name.as_str() == class_name) else {
        return None;
    };

    for base in &class_def.bases {
        if base.is_attribute_expr() {
            let Some(call_path) = model.resolve_call_path(base) else {
                continue;
            };
            bases.push(call_path.join("."));
            continue;
        };

        let id = match base {
            Expr::Name(expr) => expr.id.as_str(),
            _ => continue,
        };

        let Some(class_def) =  class_definitions.iter().find(|c| c.name.as_str() == id) else {
            continue;
        };

        match method_resolution_order(model, class_def.name.to_string()) {
            Some(mros) => bases_mro.push(mros),
            None => return None,
        }

        bases.push(class_def.name.to_string());
    }

    if bases_mro.iter().any(|b| has_duplicates(b)) || has_duplicates(&bases) {
        // Probably there's an inconsistent hierarchy. There must be
        // something wrong going on, don't try to figure it out.
        return None;
    }

    clean_typing_generic_mro(&mut bases, &mut bases_mro);
    let mut unmerged_mro = vec![vec![class_name]];
    unmerged_mro.extend(bases_mro.into_iter());
    unmerged_mro.push(bases);

    merge_mro(unmerged_mro)
}

/// Tries to find the next base in the MRO
fn find_next_base(bases: &mut [Vec<String>]) -> Option<String> {
    for i in 0..bases.len() {
        let head = match bases[i].get(0) {
            Some(v) => v.clone(),
            None => continue,
        };
        if !bases.iter().any(|m| m.iter().skip(1).any(|s| s == &head)) {
            for item in bases.iter_mut() {
                let Some(first) = item.get(0) else {
                    continue;
                };
                if first == &head {
                    item.remove(0);
                }
            }
            return Some(head);
        }
    }
    None
}

/// Merge the MROs in a single vector
fn merge_mro(mut mros: Vec<Vec<String>>) -> Option<Vec<String>> {
    let mut result = vec![];
    while !mros.is_empty() {
        let Some(head) = find_next_base(&mut mros) else {
            return None;
        };
        result.push(head);
        mros.retain(|s| !s.is_empty());
    }
    Some(result)
}

/// Returns true if `class_def` has at least an abstract method
fn has_abstract_method(model: &SemanticModel, class_def: &StmtClassDef) -> bool {
    let Some(mro) = method_resolution_order(model, class_def.name.to_string()) else {
        return false;
    };
    let class_definitions = class_definitions(model).collect_vec();
    let mut abstract_methods = HashSet::new();
    for ancestor in mro.iter().rev() {
        // There could be multiple definition of a class, maybe hidden behind a conditional.
        // Here we get all the bodies of all classes with that name and merge them into
        // a single vector.
        // Merging them doesn't cause any resolution issues, multiple definitions of a function
        // won't clash in any case.
        let bodies = class_definitions
            .iter()
            .filter(|c| c.name.as_str() == ancestor)
            .flat_map(|c| &c.body)
            .collect_vec();

        for stmt in bodies {
            // Get the the name of the function
            let names = match stmt {
                Stmt::FunctionDef(_) | Stmt::AsyncFunctionDef(_) => vec![cast::name(stmt)],
                Stmt::AnnAssign(a) => vec![a.target.as_name_expr().unwrap().id.as_str()],
                // Assignment staments can have multiple targets (e.g. a, b = 1, 2), so we need to
                // get all of them in case there is an assignment to a inherited abstract method that
                // is overridden in the current class.
                Stmt::Assign(a) => a
                    .targets
                    .iter()
                    .map(|t| t.as_name_expr().unwrap().id.as_str())
                    .collect_vec(),
                _ => continue,
            };

            let decorators = match stmt {
                Stmt::FunctionDef(_) | Stmt::AsyncFunctionDef(_) => cast::decorator_list(stmt),
                _ => &[],
            };

            if decorators.is_empty() {
                // If the current function doesn't have decorators, we can safely assume that
                // it's not abstract. If it was found as abstract in an ancestor it's now overridden.
                // Also we remove all names as there could be multiple assignments to different functions.
                // (e.g. a, b = foo, bar)
                for name in &names {
                    abstract_methods.remove(name);
                }
            }

            for decorator in decorators {
                let Some(call_path) = model.resolve_call_path(&decorator.expression) else {
                    continue;
                };
                if ABC_METHODS.contains(&call_path.join(".").as_str()) {
                    // We get only the first name as we reach this point only if at least
                    // a decorator is present, so we can safely assume that this is a
                    // function definition.
                    abstract_methods.insert(names[0]);
                }
            }
        }
    }
    !abstract_methods.is_empty()
}

/// Get all the class definitions in `model`
fn class_definitions<'a>(model: &'a SemanticModel) -> impl Iterator<Item = &'a StmtClassDef> {
    fn filter_map_statement<'b>(def: &'b Definition) -> Option<&'b Stmt> {
        let Definition::Member(Member{
            kind: MemberKind::Class | MemberKind::NestedClass,
            stmt, ..
        }) = def  else {
            return None;
        };
        Some(*stmt)
    }
    fn filter_map_class(stmt: &Stmt) -> Option<&StmtClassDef> {
        match stmt {
            Stmt::ClassDef(def) => Some(def),
            _ => None,
        }
    }
    model
        .definitions
        .iter()
        .filter_map(filter_map_statement)
        .filter_map(filter_map_class)
}

/// Returns true if `class_def` has an abstract base
fn has_abstract_metaclass(model: &SemanticModel, class_def: &StmtClassDef) -> bool {
    for keyword in &class_def.keywords {
        let Some(id)  = &keyword.arg else {
            continue;
        };

        if id.as_str() != "metaclass" {
            continue;
        }

        let Some(call_path) = model.resolve_call_path(&keyword.value) else {
            continue;
        };

        if ABC_METACLASSES.contains(&call_path.join(".").as_str()) {
            return true;
        }
    }

    // No metaclass found, let's check the ancestors
    ancestors(model, class_def)
        .iter()
        .any(|item| has_abstract_metaclass(model, item.1))
}

/// Returns a map of all the ancestors of `class_def`
fn ancestors<'a>(
    model: &'a SemanticModel,
    class_def: &'a StmtClassDef,
) -> HashMap<&'a str, &'a StmtClassDef> {
    let mut visited: HashMap<&'a str, &'a StmtClassDef> = HashMap::new();
    let class_definitions = class_definitions(model).collect_vec();

    for base in &class_def.bases {
        let id = match base {
            Expr::Name(expr) => expr.id.as_str(),
            Expr::Attribute(expr) => expr.attr.as_str(),
            _ => continue,
        };

        if visited.get(id).is_some() {
            continue;
        }

        let Some(class_def) =  class_definitions.iter().find(|class_def| class_def.name.as_str() == id) else {
            continue;
        };

        visited.insert(id, class_def);
        visited.extend(ancestors(model, class_def));
    }

    visited
}

/// PLE0110
pub(crate) fn abstract_class_instantiated(checker: &mut Checker, expr: &Expr, func: &Expr) {
    if !expr.is_call_expr() {
        return;
    }

    let Expr::Name(ExprName{ id: func_name, ..}) = func else {
        return;
    };

    let definitions = class_definitions(checker.semantic()).collect_vec();
    for class_def in &definitions {
        if *func_name != class_def.name.as_str() {
            // This class is not being instantiated, skip it
            continue;
        }

        if !has_abstract_method(checker.semantic(), class_def) {
            // If the class doesn't have abstract methods it can be instatiated
            continue;
        }

        if has_abstract_metaclass(checker.semantic(), class_def) {
            checker.diagnostics.push(Diagnostic::new(
                AbstractClassInstantiated {
                    class_name: String::from(class_def.name.clone()),
                },
                expr.range(),
            ));
            return;
        }
    }
}
