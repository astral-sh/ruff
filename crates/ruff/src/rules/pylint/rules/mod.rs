pub use assert_on_string_literal::{assert_on_string_literal, AssertOnStringLiteral};
pub use await_outside_async::{await_outside_async, AwaitOutsideAsync};
pub use bad_str_strip_call::{bad_str_strip_call, BadStrStripCall};
pub use bad_string_format_type::{bad_string_format_type, BadStringFormatType};
pub use bidirectional_unicode::{bidirectional_unicode, BidirectionalUnicode};
pub use binary_op_exception::{binary_op_exception, BinaryOpException};
pub use collapsible_else_if::{collapsible_else_if, CollapsibleElseIf};
pub use compare_to_empty_string::{compare_to_empty_string, CompareToEmptyString};
pub use comparison_of_constant::{comparison_of_constant, ComparisonOfConstant};
pub use continue_in_finally::{continue_in_finally, ContinueInFinally};
pub use global_statement::{global_statement, GlobalStatement};
pub use global_variable_not_assigned::GlobalVariableNotAssigned;
pub use invalid_all_format::{invalid_all_format, InvalidAllFormat};
pub use invalid_all_object::{invalid_all_object, InvalidAllObject};
pub use invalid_envvar_default::{invalid_envvar_default, InvalidEnvvarDefault};
pub use invalid_envvar_value::{invalid_envvar_value, InvalidEnvvarValue};
pub use invalid_string_characters::{
    invalid_string_characters, InvalidCharacterBackspace, InvalidCharacterEsc, InvalidCharacterNul,
    InvalidCharacterSub, InvalidCharacterZeroWidthSpace,
};
pub use load_before_global_declaration::{
    load_before_global_declaration, LoadBeforeGlobalDeclaration,
};
pub use logging::{logging_call, LoggingTooFewArgs, LoggingTooManyArgs};
pub use magic_value_comparison::{magic_value_comparison, MagicValueComparison};
pub use manual_import_from::{manual_from_import, ManualFromImport};
pub use nonlocal_without_binding::NonlocalWithoutBinding;
pub use property_with_parameters::{property_with_parameters, PropertyWithParameters};
pub use redefined_loop_name::{redefined_loop_name, RedefinedLoopName};
pub use repeated_isinstance_calls::{repeated_isinstance_calls, RepeatedIsinstanceCalls};
pub use return_in_init::{return_in_init, ReturnInInit};
pub use sys_exit_alias::{sys_exit_alias, SysExitAlias};
pub use too_many_arguments::{too_many_arguments, TooManyArguments};
pub use too_many_branches::{too_many_branches, TooManyBranches};
pub use too_many_return_statements::{too_many_return_statements, TooManyReturnStatements};
pub use too_many_statements::{too_many_statements, TooManyStatements};
pub use unnecessary_direct_lambda_call::{
    unnecessary_direct_lambda_call, UnnecessaryDirectLambdaCall,
};
pub use useless_else_on_loop::{useless_else_on_loop, UselessElseOnLoop};
pub use useless_import_alias::{useless_import_alias, UselessImportAlias};
pub use useless_return::{useless_return, UselessReturn};
pub use yield_in_init::{yield_in_init, YieldInInit};

mod assert_on_string_literal;
mod await_outside_async;
mod bad_str_strip_call;
mod bad_string_format_type;
mod bidirectional_unicode;
mod binary_op_exception;
mod collapsible_else_if;
mod compare_to_empty_string;
mod comparison_of_constant;
mod continue_in_finally;
mod global_statement;
mod global_variable_not_assigned;
mod invalid_all_format;
mod invalid_all_object;
mod invalid_envvar_default;
mod invalid_envvar_value;
mod invalid_string_characters;
mod load_before_global_declaration;
mod logging;
mod magic_value_comparison;
mod manual_import_from;
mod nonlocal_without_binding;
mod property_with_parameters;
mod redefined_loop_name;
mod repeated_isinstance_calls;
mod return_in_init;
mod sys_exit_alias;
mod too_many_arguments;
mod too_many_branches;
mod too_many_return_statements;
mod too_many_statements;
mod unnecessary_direct_lambda_call;
mod useless_else_on_loop;
mod useless_import_alias;
mod useless_return;
mod yield_in_init;
