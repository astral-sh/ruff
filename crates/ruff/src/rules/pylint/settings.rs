//! Settings for the `pylint` plugin.

use anyhow::anyhow;
use serde::{Deserialize, Serialize};

use ruff_macros::CacheKey;
use ruff_python_ast::Constant;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize, CacheKey)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
pub enum ConstantType {
    Bytes,
    Complex,
    Float,
    Int,
    Str,
    Tuple,
}

impl TryFrom<&Constant> for ConstantType {
    type Error = anyhow::Error;

    fn try_from(value: &Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::Bytes(..) => Ok(Self::Bytes),
            Constant::Complex { .. } => Ok(Self::Complex),
            Constant::Float(..) => Ok(Self::Float),
            Constant::Int(..) => Ok(Self::Int),
            Constant::Str(..) => Ok(Self::Str),
            Constant::Bool(..) | Constant::Ellipsis | Constant::None => {
                Err(anyhow!("Singleton constants are unsupported"))
            }
        }
    }
}

#[derive(Debug, CacheKey)]
pub struct Settings {
    pub allow_magic_value_types: Vec<ConstantType>,
    pub max_args: usize,
    pub max_returns: usize,
    pub max_branches: usize,
    pub max_statements: usize,
    pub max_public_methods: usize,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            allow_magic_value_types: vec![ConstantType::Str, ConstantType::Bytes],
            max_args: 5,
            max_returns: 6,
            max_branches: 12,
            max_statements: 50,
            max_public_methods: 20,
        }
    }
}
