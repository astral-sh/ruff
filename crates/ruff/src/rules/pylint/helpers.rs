use std::fmt;

use rustpython_parser::ast;
use rustpython_parser::ast::{CmpOp, Constant, Expr, Keyword};

use ruff_python_semantic::analyze::function_type;
use ruff_python_semantic::{ScopeKind, SemanticModel};

use crate::settings::Settings;

/// Returns the value of the `name` parameter to, e.g., a `TypeVar` constructor.
pub(super) fn type_param_name<'a>(args: &'a [Expr], keywords: &'a [Keyword]) -> Option<&'a str> {
    // Handle both `TypeVar("T")` and `TypeVar(name="T")`.
    let name_param = keywords
        .iter()
        .find(|keyword| {
            keyword
                .arg
                .as_ref()
                .map_or(false, |keyword| keyword.as_str() == "name")
        })
        .map(|keyword| &keyword.value)
        .or_else(|| args.get(0))?;
    if let Expr::Constant(ast::ExprConstant {
        value: Constant::Str(name),
        ..
    }) = &name_param
    {
        Some(name)
    } else {
        None
    }
}

pub(super) fn in_dunder_init(semantic: &SemanticModel, settings: &Settings) -> bool {
    let scope = semantic.scope();
    let (ScopeKind::Function(ast::StmtFunctionDef {
        name,
        decorator_list,
        ..
    })
    | ScopeKind::AsyncFunction(ast::StmtAsyncFunctionDef {
        name,
        decorator_list,
        ..
    })) = scope.kind
    else {
        return false;
    };
    if name != "__init__" {
        return false;
    }
    let Some(parent) = scope.parent.map(|scope_id| &semantic.scopes[scope_id]) else {
        return false;
    };

    if !matches!(
        function_type::classify(
            name,
            decorator_list,
            parent,
            semantic,
            &settings.pep8_naming.classmethod_decorators,
            &settings.pep8_naming.staticmethod_decorators,
        ),
        function_type::FunctionType::Method
    ) {
        return false;
    }
    true
}

/// A wrapper around [`CmpOp`] that implements `Display`.
#[derive(Debug)]
pub(super) struct CmpOpExt(CmpOp);

impl From<&CmpOp> for CmpOpExt {
    fn from(cmp_op: &CmpOp) -> Self {
        CmpOpExt(*cmp_op)
    }
}

impl fmt::Display for CmpOpExt {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let representation = match self.0 {
            CmpOp::Eq => "==",
            CmpOp::NotEq => "!=",
            CmpOp::Lt => "<",
            CmpOp::LtE => "<=",
            CmpOp::Gt => ">",
            CmpOp::GtE => ">=",
            CmpOp::Is => "is",
            CmpOp::IsNot => "is not",
            CmpOp::In => "in",
            CmpOp::NotIn => "not in",
        };
        write!(f, "{representation}")
    }
}
