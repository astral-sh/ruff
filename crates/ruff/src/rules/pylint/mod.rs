//! Rules from [Pylint](https://pypi.org/project/pylint/).
mod helpers;
pub(crate) mod rules;
pub mod settings;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use crate::assert_messages;
    use anyhow::Result;

    use regex::Regex;
    use test_case::test_case;

    use crate::registry::Rule;
    use crate::rules::pylint;
    use crate::settings::types::PythonVersion;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::AwaitOutsideAsync, Path::new("await_outside_async.py"); "PLE1142")]
    #[test_case(Rule::AssertOnStringLiteral, Path::new("assert_on_string_literal.py"); "PLW0129")]
    #[test_case(Rule::BadStrStripCall, Path::new("bad_str_strip_call.py"); "PLE01310")]
    #[test_case(Rule::BadStringFormatType, Path::new("bad_string_format_type.py"); "PLE1307")]
    #[test_case(Rule::BidirectionalUnicode, Path::new("bidirectional_unicode.py"); "PLE2502")]
    #[test_case(Rule::BinaryOpException, Path::new("binary_op_exception.py"); "PLW0711")]
    #[test_case(Rule::CollapsibleElseIf, Path::new("collapsible_else_if.py"); "PLR5501")]
    #[test_case(Rule::CompareToEmptyString, Path::new("compare_to_empty_string.py"); "PLC1901")]
    #[test_case(Rule::ComparisonOfConstant, Path::new("comparison_of_constant.py"); "PLR0133")]
    #[test_case(Rule::RepeatedIsinstanceCalls, Path::new("repeated_isinstance_calls.py"); "PLR1701")]
    #[test_case(Rule::ManualFromImport, Path::new("import_aliasing.py"); "PLR0402")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_0.py"); "PLR1722_0")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_1.py"); "PLR1722_1")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_2.py"); "PLR1722_2")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_3.py"); "PLR1722_3")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_4.py"); "PLR1722_4")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_5.py"); "PLR1722_5")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_6.py"); "PLR1722_6")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_7.py"); "PLR1722_7")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_8.py"); "PLR1722_8")]
    #[test_case(Rule::SysExitAlias, Path::new("sys_exit_alias_9.py"); "PLR1722_9")]
    #[test_case(Rule::ContinueInFinally, Path::new("continue_in_finally.py"); "PLE0116")]
    #[test_case(Rule::GlobalStatement, Path::new("global_statement.py"); "PLW0603")]
    #[test_case(Rule::GlobalVariableNotAssigned, Path::new("global_variable_not_assigned.py"); "PLW0602")]
    #[test_case(Rule::ImportSelf, Path::new("import_self/module.py"); "PLW0406")]
    #[test_case(Rule::InvalidAllFormat, Path::new("invalid_all_format.py"); "PLE0605")]
    #[test_case(Rule::InvalidAllObject, Path::new("invalid_all_object.py"); "PLE0604")]
    #[test_case(Rule::InvalidCharacterBackspace, Path::new("invalid_characters.py"); "PLE2510")]
    #[test_case(Rule::InvalidCharacterEsc, Path::new("invalid_characters.py"); "PLE2513")]
    #[test_case(Rule::InvalidCharacterNul, Path::new("invalid_characters.py"); "PLE2514")]
    #[test_case(Rule::InvalidCharacterSub, Path::new("invalid_characters.py"); "PLE2512")]
    #[test_case(Rule::InvalidCharacterZeroWidthSpace, Path::new("invalid_characters.py"); "PLE2515")]
    #[test_case(Rule::InvalidEnvvarDefault, Path::new("invalid_envvar_default.py"); "PLW1508")]
    #[test_case(Rule::InvalidEnvvarValue, Path::new("invalid_envvar_value.py"); "PLE1507")]
    #[test_case(Rule::LoggingTooFewArgs, Path::new("logging_too_few_args.py"); "PLE1206")]
    #[test_case(Rule::LoggingTooManyArgs, Path::new("logging_too_many_args.py"); "PLE1205")]
    #[test_case(Rule::MagicValueComparison, Path::new("magic_value_comparison.py"); "PLR2004")]
    #[test_case(Rule::NonlocalWithoutBinding, Path::new("nonlocal_without_binding.py"); "PLE0117")]
    #[test_case(Rule::PropertyWithParameters, Path::new("property_with_parameters.py"); "PLR0206")]
    #[test_case(Rule::RedefinedLoopName, Path::new("redefined_loop_name.py"); "PLW2901")]
    #[test_case(Rule::ReturnInInit, Path::new("return_in_init.py"); "PLE0101")]
    #[test_case(Rule::TooManyArguments, Path::new("too_many_arguments.py"); "PLR0913")]
    #[test_case(Rule::TooManyBranches, Path::new("too_many_branches.py"); "PLR0912")]
    #[test_case(Rule::TooManyReturnStatements, Path::new("too_many_return_statements.py"); "PLR0911")]
    #[test_case(Rule::TooManyStatements, Path::new("too_many_statements.py"); "PLR0915")]
    #[test_case(Rule::UnexpectedSpecialMethodSignature, Path::new("unexpected_special_method_signature.py"); "PLE0302")]
    #[test_case(Rule::UnnecessaryDirectLambdaCall, Path::new("unnecessary_direct_lambda_call.py"); "PLC3002")]
    #[test_case(Rule::LoadBeforeGlobalDeclaration, Path::new("load_before_global_declaration.py"); "PLE0118")]
    #[test_case(Rule::UselessElseOnLoop, Path::new("useless_else_on_loop.py"); "PLW0120")]
    #[test_case(Rule::UselessImportAlias, Path::new("import_aliasing.py"); "PLC0414")]
    #[test_case(Rule::UselessReturn, Path::new("useless_return.py"); "PLR1711")]
    #[test_case(Rule::YieldInInit, Path::new("yield_in_init.py"); "PLE0100")]
    #[test_case(Rule::NestedMinMax, Path::new("nested_min_max.py"); "PLW3301")]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("pylint").join(path).as_path(),
            &Settings::for_rules(vec![rule_code]),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }

    #[test]
    fn continue_in_finally() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/continue_in_finally.py"),
            &Settings {
                target_version: PythonVersion::Py37,
                ..Settings::for_rules(vec![Rule::ContinueInFinally])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn allow_magic_value_types() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/magic_value_comparison.py"),
            &Settings {
                pylint: pylint::settings::Settings {
                    allow_magic_value_types: vec![pylint::settings::ConstantType::Int],
                    ..pylint::settings::Settings::default()
                },
                ..Settings::for_rules(vec![Rule::MagicValueComparison])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_args() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_arguments_params.py"),
            &Settings {
                pylint: pylint::settings::Settings {
                    max_args: 4,
                    ..pylint::settings::Settings::default()
                },
                ..Settings::for_rules(vec![Rule::TooManyArguments])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_args_with_dummy_variables() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_arguments_params.py"),
            &Settings {
                dummy_variable_rgx: Regex::new(r"skip_.*").unwrap(),
                ..Settings::for_rules(vec![Rule::TooManyArguments])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_branches() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_branches_params.py"),
            &Settings {
                pylint: pylint::settings::Settings {
                    max_branches: 1,
                    ..pylint::settings::Settings::default()
                },
                ..Settings::for_rules(vec![Rule::TooManyBranches])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_statements() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_statements_params.py"),
            &Settings {
                pylint: pylint::settings::Settings {
                    max_statements: 1,
                    ..pylint::settings::Settings::default()
                },
                ..Settings::for_rules(vec![Rule::TooManyStatements])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }

    #[test]
    fn max_return_statements() -> Result<()> {
        let diagnostics = test_path(
            Path::new("pylint/too_many_return_statements_params.py"),
            &Settings {
                pylint: pylint::settings::Settings {
                    max_returns: 1,
                    ..pylint::settings::Settings::default()
                },
                ..Settings::for_rules(vec![Rule::TooManyReturnStatements])
            },
        )?;
        assert_messages!(diagnostics);
        Ok(())
    }
}
