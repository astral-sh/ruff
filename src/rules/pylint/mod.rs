//! Rules from [Pylint](https://pypi.org/project/pylint/2.15.7/).
pub(crate) mod rules;
pub mod settings;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::linter::test_path;
    use crate::registry::Rule;
    use crate::rules::pylint;
    use crate::settings::Settings;

    #[test_case(Rule::UselessImportAlias, Path::new("import_aliasing.py"); "PLC0414")]
    #[test_case(Rule::UnnecessaryDirectLambdaCall, Path::new("unnecessary_direct_lambda_call.py"); "PLC3002")]
    #[test_case(Rule::NonlocalWithoutBinding, Path::new("nonlocal_without_binding.py"); "PLE0117")]
    #[test_case(Rule::UsedPriorGlobalDeclaration, Path::new("used_prior_global_declaration.py"); "PLE0118")]
    #[test_case(Rule::AwaitOutsideAsync, Path::new("await_outside_async.py"); "PLE1142")]
    #[test_case(Rule::ConstantComparison, Path::new("constant_comparison.py"); "PLR0133")]
    #[test_case(Rule::PropertyWithParameters, Path::new("property_with_parameters.py"); "PLR0206")]
    #[test_case(Rule::ConsiderUsingFromImport, Path::new("import_aliasing.py"); "PLR0402")]
    #[test_case(Rule::ConsiderMergingIsinstance, Path::new("consider_merging_isinstance.py"); "PLR1701")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_0.py"); "PLR1722_0")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_1.py"); "PLR1722_1")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_2.py"); "PLR1722_2")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_3.py"); "PLR1722_3")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_4.py"); "PLR1722_4")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_5.py"); "PLR1722_5")]
    #[test_case(Rule::UseSysExit, Path::new("consider_using_sys_exit_6.py"); "PLR1722_6")]
    #[test_case(Rule::MagicValueComparison, Path::new("magic_value_comparison.py"); "PLR2004")]
    #[test_case(Rule::UselessElseOnLoop, Path::new("useless_else_on_loop.py"); "PLW0120")]
    #[test_case(Rule::GlobalVariableNotAssigned, Path::new("global_variable_not_assigned.py"); "PLW0602")]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("./resources/test/fixtures/pylint")
                .join(path)
                .as_path(),
            &Settings::for_rules(vec![rule_code]),
        )?;
        insta::assert_yaml_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test]
    fn allow_magic_value_types() -> Result<()> {
        let diagnostics = test_path(
            Path::new("./resources/test/fixtures/pylint/magic_value_comparison.py"),
            &Settings {
                pylint: pylint::settings::Settings {
                    allow_magic_value_types: vec![pylint::settings::ConstantType::Int],
                },
                ..Settings::for_rules(vec![Rule::MagicValueComparison])
            },
        )?;
        insta::assert_yaml_snapshot!(diagnostics);
        Ok(())
    }
}
