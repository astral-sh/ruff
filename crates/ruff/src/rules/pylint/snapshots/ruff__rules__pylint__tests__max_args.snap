---
source: src/rules/pylint/mod.rs
expression: diagnostics
---
- kind:
    TooManyArguments:
      c_args: 6
      max_args: 4
  location:
    row: 3
    column: 0
  end_location:
    row: 4
    column: 8
  fix: ~
  parent: ~
- kind:
    TooManyArguments:
      c_args: 6
      max_args: 4
  location:
    row: 9
    column: 0
  end_location:
    row: 10
    column: 8
  fix: ~
  parent: ~

